guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+
+        // Create our response future by wrapping the future from the inner
+        // service.
+        ResponseFuture {
+            response_future,
+            sleep,
+        }
+    }
+}
+```
+
+A key point here is that Rust's futures are _lazy_. That means nothing actually
+happens until they're `await`ed or polled. So `self.inner.call(request)` will
+return immediately without actually processing the request.
+
+Next we go ahead and implement `Future` for `ResponseFuture`:
+
+```rust
+use std::{pin::Pin, future::Future};
+
+impl<F, Response, Error> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response, Error>>,
+{
+    type Output = Result<Response, Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // What to write here?
+    }
+}
+```
+
+Ideally we want to write something like this:
+
+1. First poll `self.response_future` and if its ready return the response or error it
+   resolved to.
+2. Otherwise poll `self.sleep` and if its ready return an error.
+3. If neither future is ready return `Poll::Pending`.
+
+We might try:
+
+```rust
+fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+    match self.response_future.poll(cx) {
+        Poll::Ready(result) => return Poll::Ready(result),
+        Poll::Pending => {}
+    }
+
+    todo!()
+}
+```
+
+However that gives an error like this:
+
+```
+error[E0599]: no method named `poll` found for type parameter `F` in the current scope
+  --> src/lib.rs:56:29
+   |
+56 |         match self.response_future.poll(cx) {
+   |                             ^^^^ method not found in `F`
+   |
+   = help: items from traits can only be used if the type parameter is bounded by the trait
+help: the following traits define an item `poll`, perhaps you need to restrict type parameter `F` with one of them:
+   |
+49 | impl<F: Future, Response, Error> Future for ResponseFuture<F>
+   |      ^^^^^^^^^
+
+error: aborting due to previous error
+```
+
+Unfortunately the error we get from Rust isn't very good. It tells us to add a
+`F: Future` bound even though we've already done that with `where F:
+Future<Output = Result<Response, E>>`.
+
+The real issue has to do with [`Pin`]. The full details of pinning is outside
+the scope of this guide. If you're brand new to `Pin` we recommend ["The Why,
+What, and How of Pinning in Rust"][pin] by Jon Gjengset.
+
+What Rust is trying to tell us is that we need a `Pin<&mut F>` to be able to
+call `poll`. Accessing `F` through `self.response_future` when `self` is a
+`Pin<&mut Self>` doesn't work.
+
+One solution is to make `ResponseFuture<F>` implement [`Unpin`]. If a type
+implements `Unpin` it means we can ignore `Pin` and treat a `Pin<&mut Self>` as
+just a `&mut self`, as if the `Pin` wasn't there. As `Unpin` is an auto trait
+the compiler will automatically implement it if all types contained are also
+`Unpin`, similarly to `Send` and `Sync`.
```

Comment:
another interesting call out is that implementing `Unpin` manually is safe while doing `Send`/`Sync` is `unsafe impl`. This is because `Unpin` is safe to `impl` on a struct as long as you don't use any other unsafe code to access stuff. Its not unsafe but that is where the unsaftey comes up. So if you can impl `Unpin` on your struct and not use any `unsafe` to get any of the pins out then its totally safe. Its a bit different from Send/Sync in this case.

Reply:
Yep! That is something Jon covers in his video so don't think I'll mention it here.